sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala

```diff
@@ -226,6 +226,15 @@ object GpuHashJoin extends Arm {
       case _ => false
     }
   }
+
+  /**
+    * Copied from Apache Spark, returns whether the keys can be rewritten as a packed long. If
+    * they can, we can assume that they are packed when we extract them out.
+    */
+  def canRewriteAsLongType(keys: Seq[Expression]): Boolean = {
+    keys.forall(_.dataType.isInstanceOf[IntegralType]) &&
+      keys.map(_.dataType.defaultSize).sum <= 8
+  }
```

Comment:
Please add a permalink such as https://github.com/apache/spark/blob/29eca8c87f4e8c19c0380f7c30668fd88edee573/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala#L709-L713 to the revision this snippet is taken from for easier tracking 

The implementation was changed in https://issues.apache.org/jira/browse/SPARK-38325 in 3.2.2+ for 3.2, and in 3.3+   It should be shimmed.

Reply:
So it looks like this particular logic didn't fundamentally change in the underlying PR:

It went from this:
```
    if (keys.exists(!_.dataType.isInstanceOf[IntegralType])
      || keys.map(_.dataType.defaultSize).sum > 8) {
         return keys
    }
```

To this:

```
  private def canRewriteAsLongType(keys: Seq[Expression]): Boolean = {
    // TODO: support BooleanType, DateType and TimestampType
    keys.forall(_.dataType.isInstanceOf[IntegralType]) &&
      keys.map(_.dataType.defaultSize).sum <= 8
  }
...

  if (!canRewriteAsLongType(keys)) {
     return keys;
  }
```

So the first case is the explicit negation of the second logic which has been refactored into a private helper method and the negated output is checked here. 

In that PR, the material change being made is the underlying implementation of `HashJoin.rewriteKeyExprAt(...)`, which is a method I'm actually calling here, so whatever previous behavior should be preserved across Spark versions.

In our case we have to use this logic because we don't have access to the `LongHashedRelation` that would normally be created by the broadcast (since we're broadcasting columns).

I don't really know how to shim this, because there isn't any material difference in the logic (it was only refactored) that I'm using. The underlying method signature of `rewriteKeyExprAt` didn't change so I should be able to just call that Spark method and preserve behavior. I will add the permalink.

Reply:
yes, I ended up updating my comment that there was no change in logic, no need to shim